Mainnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs

```diff
@@ -0,0 +1,639 @@
+    /// </summary>
+    public string Name
+    {
+        get => this.PersistentState.GetString(nameof(Name));
+        private set => this.PersistentState.SetString(nameof(Name), value);
+    }
+
+    /// <summary>
+    /// Symbol for non-fungible token contract
+    /// </summary>
+    public string Symbol
+    {
+        get => this.PersistentState.GetString(nameof(Symbol));
+        private set => this.PersistentState.SetString(nameof(Symbol), value);
+    }
+
+    /// <summary>
+    /// The next token index which is going to be minted
+    /// </summary>
+    private ulong NextTokenId
+    {
+        get => this.PersistentState.GetUInt64(nameof(NextTokenId));
+        set => this.PersistentState.SetUInt64(nameof(NextTokenId), value);
+    }
+
+    private string GetTokenByIndexKey(ulong index) => $"TokenByIndex:{index}";
+
+    private ulong GetTokenByIndex(ulong index) => this.PersistentState.GetUInt64(GetTokenByIndexKey(index));
+
+    private void SetTokenByIndex(ulong index, ulong token) => this.PersistentState.SetUInt64(GetTokenByIndexKey(index), token);
+
+    private void ClearTokenByIndex(ulong index) => this.PersistentState.Clear(GetTokenByIndexKey(index));
+
+    private string GetIndexByTokenKey(ulong token) => $"IndexByToken:{token}";
+
+    private ulong GetIndexByToken(ulong token) => this.PersistentState.GetUInt64(GetIndexByTokenKey(token));
+
+    private void SetIndexByToken(ulong token, ulong index) => this.PersistentState.SetUInt64(GetIndexByTokenKey(token), index);
+
+    private void ClearIndexByToken(ulong token) => this.PersistentState.Clear(GetIndexByTokenKey(token));
+
+    private string GetTokenOfOwnerByIndexKey(Address address, ulong index) => $"TokenOfOwnerByIndex:{address}:{index}";
+
+    private ulong GetTokenOfOwnerByIndex(Address address, ulong index) => this.PersistentState.GetUInt64(GetTokenOfOwnerByIndexKey(address, index));
+
+    private void SetTokenOfOwnerByIndex(Address owner, ulong index, ulong tokenId) => this.PersistentState.SetUInt64(GetTokenOfOwnerByIndexKey(owner, index), tokenId);
+
+    private void ClearTokenOfOwnerByIndex(Address owner, ulong index) => this.PersistentState.Clear(GetTokenOfOwnerByIndexKey(owner, index));
+
+    private string IndexOfOwnerByTokenKey(Address owner, ulong tokenId) => $"IndexOfOwnerByToken:{owner}:{tokenId}";
+    private ulong GetIndexOfOwnerByToken(Address owner, ulong tokenId) => this.PersistentState.GetUInt64(IndexOfOwnerByTokenKey(owner, tokenId));
+    private void SetIndexOfOwnerByToken(Address owner, ulong tokenId, ulong index) => this.PersistentState.SetUInt64(IndexOfOwnerByTokenKey(owner, tokenId), index);
+    private void ClearIndexOfOwnerByToken(Address owner, ulong tokenId) => this.PersistentState.Clear(IndexOfOwnerByTokenKey(owner, tokenId));
+    public ulong TotalSupply
+    {
+        get => this.PersistentState.GetUInt64(nameof(TotalSupply));
+        private set => this.PersistentState.SetUInt64(nameof(TotalSupply), value);
+    }
```

Comment:
Same comment as other contracts, this should be changed for the better going forward as outlined in Issue https://github.com/stratisproject/CirrusSmartContracts/issues/29

Reply:
Why `TotalSupply` should not stay as ulong ? It has enough large value for majority of use cases.

Reply:
If the goal is to support wrapped ERC tokens, ulong won't be enough to support that since ethereum has up to 18 decimal places. `18.446744073709551615` would be the max total of WERC tokens in any of Stratis contracts

Reply:
This is not type of ERC20 token. These are assets which can not be divided, multiplied, substracted...etc. It can't have decimal point as well. So each token is corresponds to an asset. 

Original version is uint256 but i don't know why it is so .

Reply:
I am making research of reason it. But i guess this type of token is not going to use on dexes

Reply:
Talked in discord but putting here for reference, I think you are right and uint64 will be ok for NFT's